src/shims/unix/foreign_items.rs

```diff
@@ -575,18 +575,26 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     _ => ThreadId::from(pid),
                 };
 
+                // The actual representation of the CpuAffinityMask is [c_ulong; _], in practice either
+                //
+                // - [u32; 32] on 32-bit platforms
+                // - [u64; 16] everywhere else
+                let chunk_size = CpuAffinityMask::chunk_size(&this.tcx.sess.target);
+
                 if this.ptr_is_null(mask)? {
                     let einval = this.eval_libc("EFAULT");
                     this.set_last_error(einval)?;
                     this.write_scalar(Scalar::from_i32(-1), dest)?;
-                } else if cpusetsize < CpuAffinityMask::CPU_MASK_BYTES.try_into().unwrap() {
-                    // Writing the affinity mask to the pointer would write out of bounds.
+                } else if cpusetsize % u64::try_from(chunk_size).unwrap() != 0 {
+                    // we only copy whole chunks of size_of::<c_ulong>()
                     let einval = this.eval_libc("EINVAL");
                     this.set_last_error(einval)?;
                     this.write_scalar(Scalar::from_i32(-1), dest)?;
                 } else if let Some(cpuset) = this.machine.thread_cpu_affinity.get(&thread_id) {
                     let cpuset = cpuset.clone();
-                    this.write_bytes_ptr(mask, cpuset.as_slice().iter().copied())?;
+                    // we only copy whole chunks of size_of::<c_ulong>()
+                    let byte_count = chunk_size * (cpuset.as_slice().len() / chunk_size);
+                    this.write_bytes_ptr(mask, cpuset.as_slice()[..byte_count].iter().copied())?;
```

Comment:
We've already checked that the size is a multiple of chunk_size... so isn't this equivalent to passing the entire `cpuset`?